### PR TITLE
WT-2307: Fix for cursor iteration bug when pages are splitting

### PR DESCRIPTION
**Pull request by keithbostic**
@agorrod, for your review/consideration, we can talk this through at your convenience.

### CODE REVIEW

**Comment by agorrod** [comment]
Hi Keith,

Well done ! I glanced at the change, but won't have a chance to do a review
today.

Would you mind switching the Jenkins stress testing joins over to run on
this branch?

Woo. Enjoy your Christmas.

Alex

**Comment by keithbostic** [comment]
@agorrod, I switched over clang-sanitizer and test/format stress builds to this branch.

File: src/btree/bt_split.c [hunk 1]
```diff
@@ -505,9 +516,21 @@ __split_root(WT_SESSION_IMPL *session, WT_PAGE *root)
...
+	 * new structures; failures during that period are handled by discarding
+	 * the memory and returning an error code, our caller knows the split
+	 * didn't happen and proceeds accordingly. Second, this function updates
+	 * the tree, and a failure in that period is catastrophic, any partial
+	 * update to the tree requires a panic, we can't recover. Third, once
+	 * the split is complete and the tree has been fully updated, we have to
+	 * ignore most errors because the split is complete and correct, callers
+	 * have to proceed accordingly.
+	 */
+	enum { WT_SPLIT_NONE, WT_SPLIT_PANIC, WT_SPLIT_IGNORE } complete;
```
Comment (sueloverso) [inline src/btree/bt_split.c]: Minor nit: I think these would be clearer as `WT_SPLIT_ERR_RETURN, WT_SPLIT_ERR_PANIC, WT_SPLIT_ERR_IGNORE`

File: src/btree/bt_split.c [hunk 2]
```diff
@@ -977,9 +996,21 @@ __split_internal(WT_SESSION_IMPL *session, WT_PAGE *parent, WT_PAGE *page)
...
```
Comment (sueloverso) [inline src/btree/bt_split.c]: Same here.

File: src/btree/bt_split.c (hunk 1 above)
Comment (keithbostic) [inline src/btree/bt_split.c]: Sure, done.

File: src/btree/bt_split.c (hunk 2 above)
Comment (keithbostic) [inline src/btree/bt_split.c]: Sure, done.

**Comment by agorrod** [comment]
Thanks @keithbostic lgtm. I pushed a tiny change for s_all, once that passes testing I'll merge into develop.

File: src/btree/bt_split.c [hunk 3]
```diff
@@ -505,9 +516,21 @@ __split_root(WT_SESSION_IMPL *session, WT_PAGE *root)
...
+	 * new structures; failures during that period are handled by discarding
+	 * the memory and returning an error code, our caller knows the split
+	 * didn't happen and proceeds accordingly. Second, this function updates
+	 * the tree, and a failure in that period is catastrophic, any partial
+	 * update to the tree requires a panic, we can't recover. Third, once
+	 * the split is complete and the tree has been fully updated, we have to
+	 * ignore most errors because the split is complete and correct, callers
+	 * have to proceed accordingly.
+	 */
+	enum { ERR_RETURN, ERR_PANIC, ERR_IGNORE } complete;
```
Comment (sueloverso) [inline src/btree/bt_split.c]: KNF?  These are listed in the order they're described in the comment and used.  Alphabetical maybe?

Comment (agorrod) [inline src/btree/bt_split.c]: Sorry @sueloverso I didn't see your comment before merging. I don't love the duplication of this enumeration on top of your concerns. I'd be happy if you want to open a new pull request that munges this code into a format you are happy with.

Comment (sueloverso) [inline src/btree/bt_split.c]: I created it at #2416

**Pull request merged**